File: Running_OpenMP/C/code.c
#include <stdio.h>
#include <omp.h>

int main()
{

  const int finval = 100000000;

  double pi_square = 0.0;

  // thread management variables
  int thread_id, thread_num;

  // accumulation variable
  double local_pi;

  // local work management
  int my_num, my_first, my_last;

#ifdef _OPENMP
  double start_time = omp_get_wtime();
#endif
    
#pragma omp parallel default(none) shared(pi_square) \
  private(thread_id, thread_num, local_pi, my_num, my_first, my_last)
  {
    thread_id  = omp_get_thread_num();
    thread_num = omp_get_num_threads();

    if (thread_id == 0)
    printf("Using %i threads.\n", thread_num);

#pragma omp barrier
    my_num   = finval/thread_num;
    my_first = 1 + thread_id * my_num;
    my_last  = (thread_id + 1) * my_num;

    local_pi = 0.0;
    
    for ( int i=my_first ; i <= my_last; i++)
      {  
	double factor = (double) i;
	local_pi += 1.0/( factor * factor );
      }

#pragma omp atomic update
    pi_square += local_pi;
  }

#ifdef _OPENMP
  double end_time = omp_get_wtime() - start_time;
#endif

  printf ( "Pi^2 = %.10f\n", pi_square*6.0);
  
#ifdef _OPENMP
  printf ( "Time taken: %f s\n", end_time);
#endif

  return 0;
}

File: Running_OpenMP/Fortran/code.f90
Program pi

  !$ use omp_lib
  
  implicit none

  integer, parameter :: finval = 100000000
  
  double precision :: pi_square = 0.0d0

  ! thread managemeent variables
  integer thread_id, thread_num

  ! accumulation variable
  double precision local_pi

  ! local work management
  integer my_num, my_first, my_last

  ! timing
  !$ double precision start_time, end_time
  
  
  double precision :: factor
  integer:: i

  !$ start_time = omp_get_wtime()

  !$omp parallel default(none) shared(pi_square) &
  !$omp    private(thread_id, thread_num, local_pi, my_num, my_first, my_last, i, factor)

  thread_id  = omp_get_thread_num()
  thread_num = omp_get_num_threads()

  my_num   = finval/thread_num
  my_first = 1 + thread_id * my_num
  my_last  = (thread_id + 1) * my_num
  
  local_pi = 0.0
  
  do i= my_first, my_last
     factor = i
     local_pi = local_pi + 1.0d0/(factor * factor)
  enddo

  !$omp atomic update
  pi_square = pi_square + local_pi

  !$omp end parallel

  !$ end_time = omp_get_wtime() - start_time
  
  print *, 'Pi**2 =', 6.0d0 * pi_square
  !$ print *, 'Time taken: ', end_time
  
end Program pi
